zenoh/src/net/routing/hat/linkstate_peer/token.rs

```diff
@@ -0,0 +1,709 @@
+    res.session_ctxs
+        .values()
+        .filter_map(|ctx| {
+            if ctx.token {
+                Some(ctx.face.clone())
+            } else {
+                None
+            }
+        })
+        .collect()
+}
+
+#[inline]
+fn remote_client_tokens(res: &Arc<Resource>, face: &Arc<FaceState>) -> bool {
+    res.session_ctxs
+        .values()
+        .any(|ctx| ctx.face.id != face.id && ctx.token)
+}
+
+#[inline]
+fn send_forget_sourced_token_to_net_childs(
+    tables: &Tables,
+    net: &Network,
+    childs: &[NodeIndex],
+    res: &Arc<Resource>,
+    src_face: Option<&Arc<FaceState>>,
+    routing_context: Option<NodeId>,
+) {
+    for child in childs {
+        if net.graph.contains_node(*child) {
+            match tables.get_face(&net.graph[*child].zid).cloned() {
+                Some(mut someface) => {
+                    if src_face.is_none() || someface.id != src_face.unwrap().id {
+                        let wire_expr = Resource::decl_key(res, &mut someface);
+
+                        someface.primitives.send_declare(RoutingContext::with_expr(
+                            Declare {
+                                interest_id: None,
+                                ext_qos: ext::QoSType::DECLARE,
+                                ext_tstamp: None,
+                                ext_nodeid: ext::NodeIdType {
+                                    node_id: routing_context.unwrap_or(0),
+                                },
+                                body: DeclareBody::UndeclareToken(UndeclareToken {
+                                    id: 0, // Sourced tokens do not use ids
+                                    ext_wire_expr: WireExprType { wire_expr },
+                                }),
+                            },
+                            res.expr(),
+                        ));
+                    }
+                }
+                None => tracing::trace!("Unable to find face for zid {}", net.graph[*child].zid),
+            }
+        }
+    }
+}
+
+fn propagate_forget_simple_token(tables: &mut Tables, res: &Arc<Resource>) {
+    for mut face in tables.faces.values().cloned() {
+        if let Some(id) = face_hat_mut!(&mut face).local_tokens.remove(res) {
+            face.primitives.send_declare(RoutingContext::with_expr(
+                Declare {
+                    interest_id: None,
+                    ext_qos: ext::QoSType::DECLARE,
+                    ext_tstamp: None,
+                    ext_nodeid: ext::NodeIdType::DEFAULT,
+                    body: DeclareBody::UndeclareToken(UndeclareToken {
+                        id,
+                        ext_wire_expr: WireExprType::null(),
+                    }),
+                },
+                res.expr(),
+            ));
+        }
+        for res in face_hat!(face)
+            .local_tokens
+            .keys()
+            .cloned()
+            .collect::<Vec<Arc<Resource>>>()
```

Comment:
Is `collect` really needed?

Reply:
Yes. Otherwise there would be a double borrow on tables.